---
title: "Deconvolution of bulk FFPEOS RNA-seq data with granulator"
output:
  BiocStyle::html_document
author:
- name: Sulev Koks
  affiliation: CMMIT, Murdoch University, Murdoch, Perth, WA.
date: "2024-04-14"
package: "`r pkg_ver('BiocStyle')`"
abstract: >
  In this study, we processed 90 FFPE samples from PathWest with an almost equal number of osteosarcoma and healthy tissues. RNA was extracted from Paraffin-embedded tissue; RNA was sequenced, the sequencing data was analysed and gene expression was compared to healthy of the same patients. Differentially expressed genes in osteosarcoma derived samples were identified and the functions of those genes were explored. This result was combined to perform meta-analysis with our previous studies based on FFPE and fresh samples.
editor_options: 
  chunk_output_type: console
---
This script is based on:
Pfister S, Kuettel V, Ferrero E (2022). _granulator: Rapid benchmarking of methods for *in
  silico* deconvolution of bulk RNA-seq data_. R package version 1.6.0,
  <https://github.com/xanibas/granulator>.

The package can be loaded using: 

```{r, eval=TRUE}
library(granulator)
```

# Data

## Bulk FFPE OS RNA-seq

Data frame and matrix files.

```{r}
load("osffpeGenes89.RData")
load("mosffpeGenes89.RData")
```

```{r}
osffpeGenes89[1:5, 1:5]
mosffpeGenes89[1:5, 1:5]
```

## Reference profiles
Data frame and matrix files.

Profiles are based on scRNAseq data from six OS patient samples (GSE162454). `Seurat` package was used to find conserved markers taht define clle clusters. 
To identify canonical cell type marker genes that are conserved across conditions, we will use the `FindConservedMarkers()` function. This function performs differential gene expression testing for each dataset/group and combines the p-values using meta-analysis methods from the `MetaDE` R package. For example, we can calculate the genes that are conserved markers irrespective of treatment condition in different clusters.

```{r}
load("OSscRNAprofile.RData")
load("matrixOSscRNAprofile.RData")
```

```{r}
OSscRNAprofile[1:5, 1:5]
mOSscRNAprofile[1:5, 1:5]
```

# Workflow

The `r Biocpkg("granulator")` workflow consists of four steps:

1. **Reference profiles** these profiles are generated from the single cell RNAseq data that have been integrated and profiles identified. Another option is using bulk RNAseq from isolated cell types like in this paper (PMID: 25545474 Reemann P, Reimann E, Ilmjärv S, Porosaar O, Silm H, Jaks V, Vasar E, Kingo K, Kõks S. Melanocytes in the skin--comparative whole transcriptome analysis of main skin cell types. PLoS One. 2014 Dec 29;9(12):e115717. doi: 10.1371/journal.pone.0115717)

2. **Deconvolution** bulk RNAseq data are overlaid by the profile data.

3. **Benchmarking** of the cell type proportions. 

4. **Correlation** cell type proportions are correlated with each other.

## Deconvolution of bulk RNA-seq data

Once suitable reference profiles have been generated, we use `deconvolute()` to estimate cell type proportions from the tissue bulk RNA-seq dataset. The function takes a matrix dataset to be deconvoluted, a matrix or a list of reference profile matrices, and a vector of deconvolution methods. 

To run the analysis, we need profile file and bulk expression file. I have two .csv files.

```{r}
library(readr)
OSscRNAprofile <- read_csv("OSscRNAprofile.csv")
View(OSscRNAprofile)

osffpeGenes90 <- read_csv("osffpeGenes90.csv")
View(osffpeGenes90)
```

We need to convert them to dataframes.

```{r}
OSscRNAprofile <- data.frame(OSscRNAprofile)
osffpeGenes90 <- data.frame(osffpeGenes90)
```

Change the rownames.

```{r}
rownames(OSscRNAprofile) <- OSscRNAprofile$gene
rownames(osffpeGenes90) <- osffpeGenes90$gene
```

And delete to column "gene"

```{r}
OSscRNAprofile$gene <- NULL
osffpeGenes90$gene <- NULL
```
```{r}
save(OSscRNAprofile, file="OSscRNAprofile.RData")
save(osffpeGenes90, file="osffpeGenes90.RData")
```

Convert into matrix

```{r}
mosffpeGenes90 <- as.matrix(osffpeGenes90)
mOSscRNAprofile <- as.matrix(OSscRNAprofile)
```

```{r}
save(mosffpeGenes90, file="matrixosffpeGenes90.RData")
save(mOSscRNAprofile , file="matrixOSscRNAprofile.RData")
```

```{r eval=FALSE}
write.csv(mosffpeGenes90, file="matrixosffpeGenes90.csv")
write.csv(mOSscRNAprofile, file="matrixOSscRNAprofile.csv")
```
```{r}
osffpeGenes20 <- osffpeGenes90[, c(1:20)]
mosffpeGenes20 <- as.matrix(osffpeGenes20)
```

Sample #42, Q16B027819YA23 because it is not variable and gives the error:
"Error in svm.default(x = df$x, y = as.vector(df$y[, z]), nu = i, type = "nu-regression",  : 
  NA/NaN/Inf in foreign function call (arg 4)"

```{r}
osffpeGenes89 <- osffpeGenes90[,-c(42)]
mosffpeGenes89 <- as.matrix(osffpeGenes89)
```

```{r}
osffpeGenes89 <- read_csv("osffpeGenes89.csv")
osffpeGenes89 <- data.frame(osffpeGenes89)
rownames(osffpeGenes89)<- osffpeGenes89$gene
osffpeGenes89$gene <- NULL
mosffpeGenes89 <- as.matrix(osffpeGenes89)
```
```{r}
write.csv(mosffpeGenes89, file="matrixosffpeGenes89.csv")
write.csv(osffpeGenes89, file="osffpeGenes89.csv")
```


```{r}
save(osffpeGenes89, file="osffpeGenes89.RData")
save(mosffpeGenes89, file="mosffpeGenes89.RData")
```

## Deconvolution

```{r, results='hide'}
decon89 <- deconvolute(m = mosffpeGenes89, sigMatrix = mOSscRNAprofile)
```

## Benchmarking

We can look at the cell type proportions computed by the different models using the `sig1` reference profile:

```{r}
decon89$proportions$dtangle_sig1[1:5, 1:5]
decon89$proportions$nnls_sig1[1:5, 1:5]
decon89$proportions$ols_sig1[1:5, 1:5]
decon89$proportions$qprog_sig1[1:5, 1:5]
decon89$proportions$qprogwc_sig1[1:5, 1:5]
decon89$proportions$rls_sig1[1:5, 1:5]
decon89$proportions$svr_sig1[1:5, 1:5]
```

We can plot the estimated cell type proportions with the function `plot_proportions()`.

```{r}
# plot cell type proportions for dtangle model on sig1 reference profile
plot_proportions(deconvoluted = decon89, method = 'dtangle', signature = 'sig1')
plot_proportions(deconvoluted = decon89, method = 'ols', signature = 'sig1')
```

The function `plot_deconvolute()` shows the estimated cell type proportions. This helps to compare different `deconvolution` methods. If the option `scale` is TRUE, the proportions are transformed into standard scores.

```{r, fig.retina = 1}
# plot cell type proportions
plot_deconvolute(deconvoluted = decon89, scale = TRUE, labels = FALSE)
```

## Correlation analysis of deconvoluted proportions

We can assess the performance of the different `deconvolution` methods by computing the correlation between estimated cell type proportions generated by all methods using the `correlate()` function.

```{r}
# correlation analysis
correl89 <- correlate(deconvoluted = decon89)
```

The `plot_correlate()` visualises the results of `correlate()` as a heatmap:

```{r}
# correlation heatmap
plot_correlate(correlated = correl89, method="heatmap", legend=TRUE)
```

The average correlations between methods clustered by cell type can be seen this way:

```{r}
# correlation mean summary statistics
head(correl89$summary)
```

The average correlations across cell types and the `deconvolution` method:

```{r}
# deconvolution method ranking
head(correl89$rank)
```

# Session Info

```{r}
# print session info
sessionInfo()
```
