---
title: "Deconvolution of bulk FFPEOS RNA-seq data with granulator"
output:
  BiocStyle::html_document
author:
- name: Sulev Koks
  affiliation: CMMIT, Murdoch University, Murdoch, Perth, WA.
date: "`r doc_date()`"
package: "`r pkg_ver('BiocStyle')`"
abstract: >
  In this study, we processed 90 FFPE samples from PathWest with an almost equal number of osteosarcoma and healthy tissues. RNA was extracted from Paraffin-embedded tissue; RNA was sequenced, the sequencing data was analysed and gene expression was compared to healthy of the same patients. Differentially expressed genes in osteosarcoma derived samples were identified and the functions of those genes were explored. This result was combined to perform meta-analysis with our previous studies based on FFPE and fresh samples.
editor_options: 
  chunk_output_type: console
---

The package can be loaded using: 

```{r, eval=TRUE}
library(granulator)
```

# Data

## Bulk FFPE OS RNA-seq

Data frame and matrix files.

```{r}
load("osffpeGenes89.RData")
load("mosffpeGenes89.RData")
```

```{r}
osffpeGenes90[1:5, 1:5]
mosffpeGenes90[1:5, 1:5]
```

## Reference profiles
Data frame and matrix files.

Profiles are based on scRNAseq data from six OS patient samples (GSE162454). `Seurat` package was used to find conserved markers taht define clle clusters. 
To identify canonical cell type marker genes that are conserved across conditions, we provide the `FindConservedMarkers()` function. This function performs differential gene expression testing for each dataset/group and combines the p-values using meta-analysis methods from the `MetaDE` R package. For example, we can calculate the genes that are conserved markers irrespective of treatment condition in different clusters.

```{r}
load("OSscRNAprofile.RData")
load("matrixOSscRNAprofile.RData")
```

```{r}
OSscRNAprofile[1:5, 1:5]
mOSscRNAprofile[1:5, 1:5]
```

# Workflow

The `r Biocpkg("granulator")` workflow consists of four steps:

1. **Reference profiles**: Reference profiles for deconvolution are usually generated by differential expression analysis on *bulk RNA-seq* generated from isolated cell types or cell-type clusters identified by *single cell RNA-seq*;

2. **Deconvolution**: Bulk RNA-seq data from heterogeneous samples is then
deconvoluted using one or more reference profiles and deconvolution methods;

3. **Benchmarking**: Estimated cell type proportions are benchmarked against 
measured cell type proportions to assess deconvolution performance. 
Measured proportions are usually generated from fluorescence-activated cell sorting or single cell RNA-seq data;

4. **Correlation**: Benchmarked reference profiles can be used to deconvolve
bulk RNA-seq data where individual cell types abundances are unknown. The 
deconvoluted cell type proportions can be correlated with each other 
in order to assess the degree of similarity in predictions across methods.

## Deconvolution of bulk RNA-seq data

Once suitable reference profiles have been generated, we use `deconvolute()` to estimate cell type proportions from the tissue bulk RNA-seq dataset. The function takes a matrix dataset to be deconvoluted, a matrix or a list of reference profile matrices, and a vector of deconvolution methods. 

All data matrices need to be normalized to TPM from raw counts with the function `get_TPM()`. By default, `deconvolute()` sequentially runs 
all methods available. Optionally, we can provide a selected list of methods and the number of available processing cores 
to minimize computation time. Every reference profile matrix is tested in combination with every selected method.

```{r}
library(readr)
OSscRNAprofile <- read_csv("OSscRNAprofile.csv")
View(OSscRNAprofile)

osffpeGenes90 <- read_csv("osffpeGenes90.csv")
View(osffpeGenes90)
```

```{r}
OSscRNAprofile <- data.frame(OSscRNAprofile)
osffpeGenes90 <- data.frame(osffpeGenes90)
```

```{r}
rownames(OSscRNAprofile) <- OSscRNAprofile$gene
rownames(osffpeGenes90) <- osffpeGenes90$gene
```
```{r}
OSscRNAprofile$gene <- NULL
osffpeGenes90$gene <- NULL
```
```{r}
save(OSscRNAprofile, file="OSscRNAprofile.RData")
save(osffpeGenes90, file="osffpeGenes90.RData")
```
```{r}
mosffpeGenes90 <- as.matrix(osffpeGenes90)
mOSscRNAprofile <- as.matrix(OSscRNAprofile)
mosffpeGenes20 <- as.matrix(osffpeGenes20)
```

```{r}
save(mosffpeGenes90, file="matrixosffpeGenes90.RData")
save(mOSscRNAprofile , file="matrixOSscRNAprofile.RData")
```

```{r eval=FALSE}
write.csv(mosffpeGenes90, file="matrixosffpeGenes90.csv")
write.csv(mOSscRNAprofile, file="matrixOSscRNAprofile.csv")
```
```{r}
osffpeGenes20 <- osffpeGenes90[, c(1:20)]
mosffpeGenes20 <- as.matrix(osffpeGenes20)
```

Sample #42, Q16B027819YA23 because it is not variable and gives the error:
"Error in svm.default(x = df$x, y = as.vector(df$y[, z]), nu = i, type = "nu-regression",  : 
  NA/NaN/Inf in foreign function call (arg 4)"

```{r}
osffpeGenes89 <- osffpeGenes90[,-c(42)]
mosffpeGenes89 <- as.matrix(osffpeGenes89)
```

```{r}
osffpeGenes89 <- read_csv("osffpeGenes89.csv")
osffpeGenes89 <- data.frame(osffpeGenes89)
rownames(osffpeGenes89)<- osffpeGenes89$gene
osffpeGenes89$gene <- NULL
mosffpeGenes89 <- as.matrix(osffpeGenes89)
```
```{r}
write.csv(mosffpeGenes89, file="matrixosffpeGenes89.csv")
write.csv(osffpeGenes89, file="osffpeGenes89.csv")
```


```{r}
save(osffpeGenes89, file="osffpeGenes89.RData")
save(mosffpeGenes89, file="mosffpeGenes89.RData")
```

```{r, results='hide'}
decon89 <- deconvolute(m = mosffpeGenes89, sigMatrix = mOSscRNAprofile)
```

For each reference profile and method combination, the function returns the estimated cell type `coefficients` and `proportions` (in percentage). Although there may be slightly negative proportions, significant negative values means that deconvolution methods fails to converge on a 
biological meaningful solution, and the reference profile matrix should be further refined.

We can look at the cell type proportions computed by the support vector regression model (`svr`) using the `sig1` reference profile:

```{r}
# print cell type proportions for svr model on ABIS_S0 reference profile
decon89$proportions$dtangle_sig1[1:5, 1:5]
decon89$proportions$nnls_sig1[1:5, 1:5]
decon89$proportions$ols_sig1[1:5, 1:5]
decon89$proportions$qprog_sig1[1:5, 1:5]
decon89$proportions$qprogwc_sig1[1:5, 1:5]
decon89$proportions$rls_sig1[1:5, 1:5]
decon89$proportions$svr_sig1[1:5, 1:5]
```

We can plot the estimated cell type proportions with the function `plot_proportions()`. Notice that while the sum of cell types proportions cannot exceed 100%, for some methods part of the bulk RNA-seq signal remains unassigned.

```{r}
# plot cell type proportions for dtangle model on sig1 reference profile
plot_proportions(deconvoluted = decon89, method = 'dtangle', signature = 'sig1')

plot_proportions(deconvoluted = decon89, method = 'ols', signature = 'sig1')
```

To plot all estimated cell type proportions we use the function `plot_deconvolute()`, which allows to compare results across deconvolution methods and cell types. The option `scale` indicates whether cell type proportions should be transformed into standard scores. Scaling is useful to directly compare deconvolution output, as the absolute percentages may vary considerably across methods.

```{r, fig.retina = 1}
# plot cell type proportions
plot_deconvolute(deconvoluted = decon89, scale = TRUE, labels = FALSE)
```

## Correlation analysis of deconvoluted proportions
When no ground truth data is available, we can assess the performance of the different deconvolution methods by computing the correlation between estimated cell type proportions generated by all methods using the `correlate()` function. By default  estimated cell type proportions are scaled to standard scores to correct for differences  in absolute estimated cell-type specific proportions across algorithms.

```{r}
# correlation analysis
correl89 <- correlate(deconvoluted = decon89)
```

The `plot_correlate()` is used to visualize the results of `correlate()`, by plotting a heatmap, where estimated cell type proportions are clustered by collinearity across cell type and deconvolution models:

```{r}
# correlation heatmap
plot_correlate(correlated = correl89, method="heatmap", legend=TRUE)
```

We observe that estimated cell type proportions are highly correlated between methods for all cell types, indicating that the deconvolution
methods agree on the assignment of cell type specific signals. The average correlations across methods by cell type can be obtained as follows:

```{r}
# correlation mean summary statistics
head(correl89$summary)
```

Of particular use is also the computation of average correlations across cell types by method, which illustrate which methods are reporting similar estimated cell type proportions:

```{r}
# deconvolution method ranking
head(correl89$rank)
```

For subsequent analysis, the estimated cell-type proportions can be now included in a linear model as covariates to account for cell type heterogeneity, or to impute cell-type specific gene expression profiles.

# Session Info

Here is the output of `sessionInfo()` on the system on which this document was compiled:

```{r}
# print session info
sessionInfo()
```

# References
