---
title: "OS scRNA-seq integration"
author: "Sulev Koks"
date: "2024-04-14"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Load libraries

```{r data}
library(Seurat)
library(patchwork)
library(magrittr)
```

# Perform analysis without integration
We have scRNAseq data from six subjects. OS1_merged.RData is an object with these six datasets merged.
We can first analyze the dataset without integration. The resulting clusters are defined both by cell type and stimulation condition, which creates challenges for downstream analysis.
run standard analysis workflow

```{r seurat-standard-workflow-2}
load("OS1_6merged.RData")
OS1_6merged
Layers(OS1_6merged)
```

```{r}
OS1_6merged <- NormalizeData(OS1_6merged)
OS1_6merged <- FindVariableFeatures(OS1_6merged)
OS1_6merged <- ScaleData(OS1_6merged)
OS1_6merged <- RunPCA(OS1_6merged)
OS1_6merged <- FindNeighbors(OS1_6merged, dims = 1:30, reduction = "pca")
OS1_6merged <- FindClusters(OS1_6merged, resolution = 0.4, cluster.name = "unintegrated_clusters")
OS1_6merged <- RunUMAP(OS1_6merged, dims = 1:30, reduction = 'pca', reduction.name = 'umap.unintegrated')
```

```{r}
DimPlot(OS1_6merged, reduction = 'umap.unintegrated', group.by=c('seurat_clusters'))
```

```{r seurat-standard-workflow}
OS1_6merged <- SCTransform(object = OS1_6merged)
OS1_6merged <- RunPCA(object = OS1_6merged)
OS1_6merged <- FindNeighbors(object = OS1_6merged, dims = 1:30)
OS1_6merged <- FindClusters(object = OS1_6merged, resolution = 0.15, cluster.name = "unintegrated_SCTclusters")
OS1_6merged <- RunUMAP(object = OS1_6merged, dims = 1:30)
DimPlot(object = OS1_6merged, reduction = 'umap')
```

```{r chaincmds,eval=FALSE}
OS1_6merged <- SCTransform(OS1_6merged) %>% RunPCA() %>% FindNeighbors(dims=1:30) %>% 
  FindClusters(resolution = 0.4) %>% RunUMAP(dims=1:30)
DimPlot(object = sce1f, reduction = 'umap')
```

# Perform integration

We now aim to integrate data from the two conditions, so that cells from the same cell type/subpopulation will cluster together. 

We often refer to this procedure as intergration/alignment. When aligning two genome sequences together, identification of shared/homologous regions can help to interpret differences between the sequences as well. Similarly for scRNA-seq integration, our goal is not to remove biological differences across conditions, but to learn shared cell types/states in an initial step - specifically because that will enable us to compare control stimulated and control profiles for these individual cell types.

The Seurat v5 integration procedure aims to return a single dimensional reduction that captures the shared sources of variance across multiple layers, so that cells in a similar biological state will cluster. The method returns a dimensional reduction (i.e. `integrated.cca`) which can be used for visualization and unsupervised clustering analysis. For evaluating performance, we can use cell type labels that are pre-loaded in the `seurat_annotations` metadata column.

```{r}
OS1_6merged <- IntegrateLayers(
  object = OS1_6merged, method = CCAIntegration,
  orig.reduction = "pca", new.reduction = 'integrated.cca',
  verbose = FALSE)

# re-join layers after integration
OS1_6merged[["RNA"]] <- JoinLayers(OS1_6merged[["RNA"]])

OS1_6merged <- FindNeighbors(OS1_6merged, reduction = "integrated.cca", dims = 1:30)
OS1_6merged <- FindClusters(OS1_6merged, resolution = 0.15)
OS1_6merged <- RunUMAP(OS1_6merged,dims = 1:30,reduction = 'integrated.cca')
```


```{r}
GetAssayData(object = OS1_6merged[["RNA"]], layer = "scale.data")[1:5,1:5]
```
as.matrix(mca@assays$RNA@data)
as.matrix(mca@assays$RNA@scale.data)
Visualization

```{r}
DimPlot(OS1_6merged, reduction = 'umap', group.by=c("seurat_clusters"))
```

Get normalised data by clusters
```{r}
OS1_6merged$seurat_clusters
```

```{r}
OS1_6merged.markers <- FindAllMarkers(OS1_6merged, only.pos = TRUE)
OS1_6merged.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)
```

```{r}
cluster1.markers <- FindMarkers(OS1_6merged, ident.1 = 1)

OS1_6merged.ROCmarkers <- FindAllMarkers(OS1_6merged, test.use = "roc", only.pos = TRUE)
OS1_6merged.ROCmarkers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)
```

```{r eval=FALSE}
write.csv(OS1_6merged.ROCmarkers, file="OS1_6merged.ALLmarkersROC.Integration.csv")
```

```{r eval=FALSE}
write.csv(OS1_6merged.markers, file="OS1_6merged.ALLmarkers.Integration.csv")
```

```{r}
cluster0.markers <- FindMarkers(OS1_6merged, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
```

```{r eval=FALSE}
write.csv(OS1_6merged.markers, file="OS1_6merged.ALLmarkers.Integration.csv")
```
```{r}
OSclusters<-split(OS1_6merged.markers, OS1_6merged.markers$cluster)
```
```{r}
OScluster0 <- OSclusters[["0"]]
```

```{r}
for (i in 1:length(OSclusters)) {
  assign(paste0("OScluster", i), as.data.frame(OSclusters[[i]]))
}
```

Save cluster as csv files

```{r}
library(readr)
file_out <- paste0("OSROCcluster", names(OSclusters), ".csv")
for(i in 1:length(OSclusters)) {
  write_csv(OSclusters[[i]], file_out[i])
}
```

```{r}
library(tidyverse)
walk2(OSclusters, file_out, ~write_csv(.x, .y))
OSclusters %>% 
  group_split(y) %>% 
  walk(~write_csv(.x, paste0(.x$y[1], ".csv")))
```

```{r}
OSROCclusters<-split(OS1_6merged.ROCmarkers, OS1_6merged.ROCmarkers$cluster)
```
```{r}
OSROCcluster0 <- OSROCclusters[["0"]]
```

```{r}
for (i in 1:length(OSROCclusters)) {
  assign(paste0("OSROCcluster", i), as.data.frame(OSROCclusters[[i]]))
}
```
```{r}
file_out <- paste0("OSROCclusters", names(OSROCclusters), ".csv")
for(i in 1:length(OSROCclusters)) {
  write_csv(OSROCclusters[[i]], file_out[i])
}
```

DefaultAssay(OS1_6merged) <- "integrated"

## Conversion of a seurat object into the single cell experiment object.

```{r}
library(scater)
library(loomR)
```

```{r}
OS16merged.sce <- as.SingleCellExperiment(OS1_6merged)
```

```{r}
p1 <- plotExpression(OS16merged.sce, features = "IBSP", x = "seurat_annotations") + theme(axis.text.x = element_text(angle = 45, 
    hjust = 1))
p2 <- plotPCA(OS16merged.sce, colour_by = "seurat_annotations")
p1 + p2
```

```{r}
p1 <- plotExpression(OS16merged.sce, features = "IFITM5", x = "seurat_annotations") + theme(axis.text.x = element_text(angle = 45, 
    hjust = 1))
p2 <- plotExpression(OS16merged.sce, features = "IFITM3", x = "seurat_annotations") + theme(axis.text.x = element_text(angle = 45, 
    hjust = 1))
p1 + p2
```



```{r}
str(OS16merged.sce$ident)
levels(OS16merged.sce$ident) <- c("Cluster 1","Cluster 2","Cluster 3","Cluster 4","Cluster 5","Cluster 6","Cluster 7","Cluster 8","Cluster 9","Cluster 10","Cluster 11")

levels(OS16merged.sce$ident)
```


```{r}
p1 <- plotExpression(OS16merged.sce, features = "CPE", x = "ident") + theme(axis.text.x = element_text(angle = 45, 
    hjust = 1))
p2 <- plotPCA(OS16merged.sce, colour_by = "ident")
p1 + p2
```



```{r}
OS1_6merged[["percent.mt"]]  <- PercentageFeatureSet(OS1_6merged, pattern = "^MT-")
OS1_6merged[["percent.rbp"]] <- PercentageFeatureSet(OS1_6merged, pattern = "^RP[SL]")

VlnPlot(OS1_6merged, features = c("nFeature_RNA","nCount_RNA","percent.mt","percent.rbp"), ncol = 4)
```

To visualize the two conditions side-by-side, we can use the `split.by` argument to show each condition colored by cluster. 
Add annotation to the meta data
```{r}
OS1_6merged@meta.data[["seurat_annotations"]] <- OS1_6merged@meta.data[["seurat_clusters"]]
levels(OS1_6merged@meta.data[["seurat_annotations"]]) <- c("Cluster 1","Cluster 2","Cluster 3","Cluster 4","Cluster 5","Cluster 6","Cluster 7","Cluster 8","Cluster 9","Cluster 10","Cluster 11")
OS1_6merged$seurat_annotations
```


```{r}
GetAssayData(object = OS1_6merged[["RNA"]], layer = "scale.data")[1:5,1:5]
```
as.matrix(mca@assays$RNA@data)
as.matrix(mca@assays$RNA@scale.data)
Visualization


```{r split.dim}
DimPlot(OS1_6merged, reduction = "umap", split.by = "seurat_clusters")
```


```{r}
print(OS1_6merged[["pca"]], dims = 1:5, nfeatures = 5)
```
```{r}
table(OS1_6merged@active.ident, OS1_6merged@meta.data$orig.ident)
```
```{r}
DimPlot(OS1_6merged, label = T)

```


```{r}
bulk <- AggregateExpression(OS1_6merged,group.by ="seurat_annotations", return.seurat = TRUE)
Cells(bulk)
```

Matrix
```{r}
bulkEXPR <- PseudobulkExpression(object = OS1_6merged, pb.method = 'aggregate', layer = 'counts', group.by = "seurat_annotations")
```
```{r}
DimPlot(OS1_6merged, group.by = "seurat_annotations", label = T)
```

```{r}
bulkEXPR[["RNA"]]
```

# pseudobulk cells only by cell type


# Identify conserved cell type markers

To identify canonical cell type marker genes that are conserved across conditions, we provide the `FindConservedMarkers()` function. This function performs differential gene expression testing for each dataset/group and combines the p-values using meta-analysis methods from the MetaDE R package. For example, we can calculated the genes that are conserved markers irrespective of stimulation condition in cluster 6 (NK cells).  

```{r conserved.markers, warning=FALSE}
Idents(OS1_6merged) <- "orig.ident"
OS1_6merged.markers <- FindConservedMarkers(OS1_6merged, ident.1 = "OSscRNA1",  grouping.var = "seurat_clusters", verbose = FALSE)
head(OS1_6merged.markers )
```

```{r eval=FALSE}
write.csv(as.data.frame(OS1_6merged.markers), file="OS1_6merged.markers.Integration.csv")
```

```{r}
write.csv(as.data.frame(bulkEXPR[["RNA"]]), file="bulkEXPRos16int.csv")
```
```{r}
write.csv(as.character(rownames(bulkEXPR[["RNA"]])), file="ROWNAMESbulkEXPRos16int.csv")
```






str(OS16merged.sce$ident)
levels(OS16merged.sce$ident) <- c("Cluster 1","Cluster 2","Cluster 3","Cluster 4","Cluster 5","Cluster 6","Cluster 7","Cluster 8","Cluster 9","Cluster 10","Cluster 11")

You can perform these same analysis on the unsupervised clustering results (stored in `seurat_clusters`), and use these conserved markers to annotate cell types in your dataset.

The `DotPlot()` function with the `split.by` parameter can be useful for viewing conserved cell type markers across conditions, showing both the expression level and the percentage of cells in a cluster expressing any given gene. Here we plot 2-3 strong marker genes for each of our 14 clusters.

```{r}
new.cluster.ids <- c("Cluster 1","Cluster 2","Cluster 3","Cluster 4","Cluster 5","Cluster 6","Cluster 7","Cluster 8","Cluster 9","Cluster 10","Cluster 11")
names(new.cluster.ids) <- levels(sce3f)
```

```{r}
#OS1_6merged$cluster <- 
#OS1_6merged$cluster <- factor(OS1_6merged$cluster, levels = c("Cluster 1","Cluster 2","Cluster 3","Cluster 4","Cluster 5","Cluster 6","Cluster 7","Cluster 8","Cluster 9","Cluster 10","Cluster 11"))
```

sce3f <- RenameIdents(sce3f, new.cluster.ids)
DimPlot(OS1_6merged, reduction = "umap", label = TRUE, pt.size = 0.5) + NoLegend() + plot_annotation(title = "OS2 dimplot")

```{r splitdotplot, fig.height = 10}

#NEEDS TO BE FIXED AND SET ORDER CORRECTLY
Idents(ifnb) <- factor(
 Idents(ifnb),
 levels = c("pDC", "Eryth","Mk", "DC", "CD14 Mono", "CD16 Mono", "B Activated", "B", "CD8 T", "NK", "T activated", "CD4 Naive T", "CD4 Memory T"))
   
markers.to.plot <- c("CD3D","CREM","HSPH1","SELL","GIMAP5","CACYBP","GNLY","NKG7","CCL5","CD8A","MS4A1","CD79A","MIR155HG","NME1","FCGR3A","VMO1","CCL2","S100A9","HLA-DQA1","GPR183","PPBP","GNG11","HBA2","HBB","TSPAN13","IL3RA","IGJ","PRSS57")
DotPlot(ifnb, features = markers.to.plot, cols = c('blue', 'red'), dot.scale = 8, split.by = "stim") + RotatedAxis()
```

# Identify differential expressed genes across conditions

Now that we've aligned the stimulated and control cells, we can start to do comparative analyses and look at the differences induced by stimulation. 

We can aggregate cells of a similar type and condition together to create "pseudobulk" profiles using the `AggregateExpression` command. As an initial exploratory analysis, we can compare pseudobulk profiles of two cell types (naive CD4 T cells, and CD14 monocytes), and compare their gene expression profiles before and after stimulation. We highlight genes that exhibit dramatic responses to interferon stimulation. As you can see, many of the same genes are upregulated in both of these cell types and likely represent a conserved interferon response pathway, though CD14 monocytes exhibit a stronger transcriptional response.


```{r scatterplots, results = 'hide', message=FALSE}
library(ggplot2)
library(cowplot)
theme_set(theme_cowplot())

aggregate_OS <- AggregateExpression(OS1_6merged,group.by = c("seurat_annotations"),return.seurat = TRUE)
genes.to.label = c("FTL", "MS4A7", "CTSBA", "CPE", "COL1A1", "COL1A2", "SPARC", "MZB1", "DERL3", "IBSP", "ALPL")

p1 <- CellScatter(aggregate_OS, "Cluster 1","Cluster 2",highlight = genes.to.label)
p2 <- LabelPoints(plot = p1, points = genes.to.label, repel = TRUE)
      
p3 <- CellScatter(aggregate_OS, "Cluster 3","Cluster 4",highlight = genes.to.label)
p4 <- LabelPoints(plot = p3, points = genes.to.label, repel = TRUE)

p2 + p4
```


We can now ask what genes change in different conditions for cells of the same type. First, we create a column in the meta.data slot to hold both the cell type and stimulation information and switch the current ident to that column. Then we use `FindMarkers()` to find the genes that are different between stimulated and control B cells. Notice that many of the top genes that show up here are the same as the ones we plotted earlier as core interferon response genes. Additionally, genes like CXCL10 which we saw were specific to monocyte and B cell interferon response show up as highly significant in this list as well. 

Please note that p-values obtained from this analysis should be interpreted with caution, as these tests treat each cell as an independent replicate, and ignore inherent correlations between cells originating from the same sample. As discussed [here](https://pubmed.ncbi.nlm.nih.gov/33257685/), DE tests across multiple conditions should expressly utilize multiple samples/replicates, and can be performed after aggregating ('pseudobulking') cells from the same sample and subpopulation together. We do not perform this analysis here, as there is a single replicate in the data, but please see our [vignette comparing healthy and diabetic samples](https://satijalab.org/seurat/articles/parsebio_sketch_integration) as an example for how to perform DE analysis across conditions.


```{r de.genes}
ifnb$celltype.stim <- paste(ifnb$seurat_annotations, ifnb$stim, sep = "_")
Idents(ifnb) <- "celltype.stim"
b.interferon.response <- FindMarkers(ifnb, ident.1 = "B_STIM", ident.2 = "B_CTRL", verbose = FALSE)
head(b.interferon.response, n = 15)
```


Another useful way to visualize these changes in gene expression is with the `split.by` option to the `FeaturePlot()` or `VlnPlot()` function. This will display FeaturePlots of the list of given genes, split by a grouping variable (stimulation condition here). Genes such as CD3D and GNLY are canonical cell type markers (for T cells and NK/CD8 T cells) that are virtually unaffected by interferon stimulation and display similar gene expression patterns in the control and stimulated group. IFI6 and ISG15, on the other hand, are core interferon response genes and are upregulated accordingly in all cell types. Finally, CD14 and CXCL10 are genes that show a cell type specific interferon response. CD14 expression decreases after stimulation in CD14 monocytes, which could lead to misclassification in a supervised analysis framework, underscoring the value of integrated analysis. CXCL10 shows a distinct upregulation in monocytes and B cells after interferon stimulation but not in other cell types. 

```{r feature.heatmaps, fig.height = 10, fig.width=10}
FeaturePlot(ifnb, features = c("CD3D", "GNLY", "IFI6"), split.by = "stim", max.cutoff = 3, cols = c("grey", "red"), reduction='umap')
```

```{r splitvln, fig.height = 12}
plots <- VlnPlot(ifnb, features = c("LYZ", "ISG15", "CXCL10"), split.by = "stim", group.by = "seurat_annotations", pt.size = 0, combine = FALSE)
wrap_plots(plots = plots, ncol = 1)
```

```{r save, include = TRUE}
```

# Perform integration with SCTransform-normalized datasets

As an alternative to log-normalization, Seurat also includes support for preprocessing of scRNA-seq using the [sctransform workflow](https://satijalab.org/seurat/articles/sctransform_vignette). The `IntegrateLayers` function also supports SCTransform-normalized data, by setting the `normalization.method` parameter, as shown below.

```{r sct, fig.height = 5}

ifnb <- LoadData('ifnb')

# split datasets and process without integration
ifnb[["RNA"]] <- split(ifnb[["RNA"]], f=ifnb$stim)
ifnb <- SCTransform(ifnb)
ifnb <- RunPCA(ifnb)
ifnb <- RunUMAP(ifnb,dims = 1:30)
DimPlot(ifnb,reduction = 'umap',group.by = c("stim","seurat_annotations"))

# integrate datasets
ifnb <- IntegrateLayers(object = ifnb, 
                           method = CCAIntegration,
                           normalization.method="SCT",
                           verbose = F)
ifnb <- FindNeighbors(ifnb,reduction = 'integrated.dr',dims = 1:30)
ifnb <- FindClusters(ifnb,resolution = 0.6)
ifnb <- RunUMAP(ifnb,dims = 1:30,reduction = 'integrated.dr')
DimPlot(ifnb,reduction = 'umap',group.by = c("stim","seurat_annotations"))


# perform differential expression
ifnb <- PrepSCTFindMarkers(ifnb)
ifnb$celltype.stim <- paste(ifnb$seurat_annotations, ifnb$stim, sep = "_")
Idents(ifnb) <- 'celltype.stim'
b.interferon.response <- FindMarkers(ifnb, ident.1 = "B_STIM", ident.2 = "B_CTRL", verbose = FALSE)

```

```

<details>
  <summary>**Session Info**</summary>
```{r}
sessionInfo()
```
</details>
