---
title: "OS scRNA-seq integration"
author: "Sulev Koks"
date: "2024-04-14"
output: html_document
editor_options: 
  chunk_output_type: console
---

This work is based on:

Hao et al. Dictionary learning for integrative, multimodal and scalable single-cell analysis.
  Nature Biotechnology (2023) [Seurat V5]

# Load libraries

```{r data}
library(Seurat)
library(patchwork)
library(magrittr)
```

# Perform analysis without integration
We have scRNAseq data from six subjects. OS1_merged.RData is an object with these six datasets merged.
We can first analyze the dataset without integration. The resulting clusters are defined both by cell type and stimulation condition, which creates challenges for downstream analysis.
run standard analysis workflow

```{r seurat-standard-workflow-2}
load("OS1_6merged.RData")
OS1_6merged
Layers(OS1_6merged)
```

```{r}
OS1_6merged <- NormalizeData(OS1_6merged)
OS1_6merged <- FindVariableFeatures(OS1_6merged)
OS1_6merged <- ScaleData(OS1_6merged)
OS1_6merged <- RunPCA(OS1_6merged)
OS1_6merged <- FindNeighbors(OS1_6merged, dims = 1:30, reduction = "pca")
OS1_6merged <- FindClusters(OS1_6merged, resolution = 0.4, cluster.name = "unintegrated_clusters")
OS1_6merged <- RunUMAP(OS1_6merged, dims = 1:30, reduction = 'pca', reduction.name = 'umap.unintegrated')
```

```{r}
DimPlot(OS1_6merged, reduction = 'umap.unintegrated', group.by=c('seurat_clusters'))
```

```{r seurat-standard-workflow}
OS1_6merged <- SCTransform(object = OS1_6merged)
OS1_6merged <- RunPCA(object = OS1_6merged)
OS1_6merged <- FindNeighbors(object = OS1_6merged, dims = 1:30)
OS1_6merged <- FindClusters(object = OS1_6merged, resolution = 0.15, cluster.name = "unintegrated_SCTclusters")
OS1_6merged <- RunUMAP(object = OS1_6merged, dims = 1:30)
DimPlot(object = OS1_6merged, reduction = 'umap')
```

```{r chaincmds,eval=FALSE}
OS1_6merged <- SCTransform(OS1_6merged) %>% RunPCA() %>% FindNeighbors(dims=1:30) %>% 
  FindClusters(resolution = 0.4) %>% RunUMAP(dims=1:30)
DimPlot(object = sce1f, reduction = 'umap')
```

# Perform integration

We now aim to integrate data from the two conditions, so that cells from the same cell type/subpopulation will cluster together. 

```{r}
OS1_6merged <- IntegrateLayers(
  object = OS1_6merged, method = CCAIntegration,
  orig.reduction = "pca", new.reduction = 'integrated.cca',
  verbose = FALSE)

# re-join layers after integration
OS1_6merged[["RNA"]] <- JoinLayers(OS1_6merged[["RNA"]])
OS1_6merged <- FindNeighbors(OS1_6merged, reduction = "integrated.cca", dims = 1:30)
OS1_6merged <- FindClusters(OS1_6merged, resolution = 0.15)
OS1_6merged <- RunUMAP(OS1_6merged,dims = 1:30,reduction = 'integrated.cca')
```

```{r}
GetAssayData(object = OS1_6merged[["RNA"]], layer = "scale.data")[1:5,1:5]
```

```{r}
DimPlot(OS1_6merged, reduction = 'umap', group.by=c("seurat_clusters"))
```

Get normalised data by clusters
```{r}
OS1_6merged$seurat_clusters
```

```{r}
OS1_6merged.markers <- FindAllMarkers(OS1_6merged, only.pos = TRUE)
OS1_6merged.markers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)
```

```{r}
cluster1.markers <- FindMarkers(OS1_6merged, ident.1 = 1)

OS1_6merged.ROCmarkers <- FindAllMarkers(OS1_6merged, test.use = "roc", only.pos = TRUE)
OS1_6merged.ROCmarkers %>%
    group_by(cluster) %>%
    dplyr::filter(avg_log2FC > 1)
```

```{r eval=FALSE}
write.csv(OS1_6merged.ROCmarkers, file="OS1_6merged.ALLmarkersROC.Integration.csv")
```

```{r eval=FALSE}
write.csv(OS1_6merged.markers, file="OS1_6merged.ALLmarkers.Integration.csv")
```

```{r}
cluster0.markers <- FindMarkers(OS1_6merged, ident.1 = 0, logfc.threshold = 0.25, test.use = "roc", only.pos = TRUE)
```

```{r eval=FALSE}
write.csv(OS1_6merged.markers, file="OS1_6merged.ALLmarkers.Integration.csv")
```
```{r}
OSclusters<-split(OS1_6merged.markers, OS1_6merged.markers$cluster)
```
```{r}
OScluster0 <- OSclusters[["0"]]
```

```{r}
for (i in 1:length(OSclusters)) {
  assign(paste0("OScluster", i), as.data.frame(OSclusters[[i]]))
}
```

Save cluster as csv files

```{r}
library(readr)
file_out <- paste0("OSROCcluster", names(OSclusters), ".csv")
for(i in 1:length(OSclusters)) {
  write_csv(OSclusters[[i]], file_out[i])
}
```

```{r}
library(tidyverse)
walk2(OSclusters, file_out, ~write_csv(.x, .y))
OSclusters %>% 
  group_split(y) %>% 
  walk(~write_csv(.x, paste0(.x$y[1], ".csv")))
```

```{r}
OSROCclusters<-split(OS1_6merged.ROCmarkers, OS1_6merged.ROCmarkers$cluster)
```
```{r}
OSROCcluster0 <- OSROCclusters[["0"]]
```

```{r}
for (i in 1:length(OSROCclusters)) {
  assign(paste0("OSROCcluster", i), as.data.frame(OSROCclusters[[i]]))
}
```
```{r}
file_out <- paste0("OSROCclusters", names(OSROCclusters), ".csv")
for(i in 1:length(OSROCclusters)) {
  write_csv(OSROCclusters[[i]], file_out[i])
}
```

DefaultAssay(OS1_6merged) <- "integrated"

## Conversion of a seurat object into the single cell experiment object.

```{r}
library(scater)
library(loomR)
```

```{r}
OS16merged.sce <- as.SingleCellExperiment(OS1_6merged)
```

```{r}
p1 <- plotExpression(OS16merged.sce, features = "IBSP", x = "seurat_annotations") + theme(axis.text.x = element_text(angle = 45, 
    hjust = 1))
p2 <- plotPCA(OS16merged.sce, colour_by = "seurat_annotations")
p1 + p2
```

```{r}
p1 <- plotExpression(OS16merged.sce, features = "IFITM5", x = "seurat_annotations") + theme(axis.text.x = element_text(angle = 45, 
    hjust = 1))
p2 <- plotExpression(OS16merged.sce, features = "IFITM3", x = "seurat_annotations") + theme(axis.text.x = element_text(angle = 45, 
    hjust = 1))
p1 + p2
```



```{r}
str(OS16merged.sce$ident)
levels(OS16merged.sce$ident) <- c("Cluster 1","Cluster 2","Cluster 3","Cluster 4","Cluster 5","Cluster 6","Cluster 7","Cluster 8","Cluster 9","Cluster 10","Cluster 11")

levels(OS16merged.sce$ident)
```

```{r}
p1 <- plotExpression(OS16merged.sce, features = "CPE", x = "ident") + theme(axis.text.x = element_text(angle = 45, 
    hjust = 1))
p2 <- plotPCA(OS16merged.sce, colour_by = "ident")
p1 + p2
```


```{r}
OS1_6merged[["percent.mt"]]  <- PercentageFeatureSet(OS1_6merged, pattern = "^MT-")
OS1_6merged[["percent.rbp"]] <- PercentageFeatureSet(OS1_6merged, pattern = "^RP[SL]")

VlnPlot(OS1_6merged, features = c("nFeature_RNA","nCount_RNA","percent.mt","percent.rbp"), ncol = 4)
```

To visualize the two conditions side-by-side, we can use the `split.by` argument to show each condition colored by cluster.  We have got 11 clusters or potential cell types.
Seurat enables adding the annotation to the meta data.

```{r}
OS1_6merged@meta.data[["seurat_annotations"]] <- OS1_6merged@meta.data[["seurat_clusters"]]
levels(OS1_6merged@meta.data[["seurat_annotations"]]) <- c("Cluster 1","Cluster 2","Cluster 3","Cluster 4","Cluster 5","Cluster 6","Cluster 7","Cluster 8","Cluster 9","Cluster 10","Cluster 11")
OS1_6merged$seurat_annotations
```

```{r}
GetAssayData(object = OS1_6merged[["RNA"]], layer = "scale.data")[1:5,1:5]
```

```{r}
DimPlot(OS1_6merged, reduction = "umap", split.by = "seurat_clusters")
```


```{r}
print(OS1_6merged[["pca"]], dims = 1:5, nfeatures = 5)
```
```{r}
table(OS1_6merged@active.ident, OS1_6merged@meta.data$orig.ident)
```
```{r}
DimPlot(OS1_6merged, label = T)

```


```{r}
bulk <- AggregateExpression(OS1_6merged,group.by ="seurat_annotations", return.seurat = TRUE)
Cells(bulk)
```

Matrix

```{r}
bulkEXPR <- PseudobulkExpression(object = OS1_6merged, pb.method = 'aggregate', layer = 'counts', group.by = "seurat_annotations")
```
```{r}
DimPlot(OS1_6merged, group.by = "seurat_annotations", label = T)
```

```{r}
bulkEXPR[["RNA"]]
```

# pseudobulk cells only by cell type

# Identify conserved cell type markers

To identify canonical cell type marker genes that are conserved across conditions, use the `FindConservedMarkers()` function. 
This function performs differential gene expression testing for each dataset/group and combines the p-values using meta-analysis methods from the MetaDE R package. This way we can define the genes that are conserved markers irrespective of other factors in the model.

```{r conserved.markers, warning=FALSE}
Idents(OS1_6merged) <- "orig.ident"
OS1_6merged.markers <- FindConservedMarkers(OS1_6merged, ident.1 = "OSscRNA1",  grouping.var = "seurat_clusters", verbose = FALSE)
head(OS1_6merged.markers )
```

```{r eval=FALSE}
write.csv(as.data.frame(OS1_6merged.markers), file="OS1_6merged.markers.Integration.csv")
```

```{r}
sessionInfo()
```

